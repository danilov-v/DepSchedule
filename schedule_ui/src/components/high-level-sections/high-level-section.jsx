import React from "react";
import { Col } from "reactstrap";

import "./high-level-sections.scss";

export function HighLevelSection({ name, length }) {
  return (
    <Col xs="auto" style={{ width: 40 * length }} className="section">
      {name}
    </Col>
  );
}
