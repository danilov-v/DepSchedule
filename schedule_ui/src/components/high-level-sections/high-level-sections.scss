.high-level-sections {
  .section {
    border: 1px solid black;
    max-width: unset;
    height: 60px;
    background: #fff;
    // color: #000;
    font-size: 1.3rem;

    &.empty {
      background: #f2f3f2;
      border: 0;

      &:last-child {
        border-right: 1px solid black;
      }
    }
  }
}
