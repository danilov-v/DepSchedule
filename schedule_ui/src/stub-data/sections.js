export const SECTIONS = [
  {
    name: "A",
    startDate: new Date(2019, 7, 16),
    endDate: new Date(2019, 8, 16),
  },
  {
    name: "B",
    startDate: new Date(2019, 8, 17),
    endDate: new Date(2019, 8, 26),
  },
  {
    name: "C",
    startDate: new Date(2019, 8, 27),
    endDate: new Date(2019, 9, 10),
  },
  {
    name: "D",
    startDate: new Date(2019, 9, 12),
    endDate: new Date(2019, 9, 16),
  },
];
