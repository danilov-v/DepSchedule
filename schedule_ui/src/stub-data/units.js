export const UNITS = [
  {
    title: "Система управления 1",
    unitLevel: 1,
    unitId: 1,
    childUnit: [
      {
        title: "Система управления 1.1",
        parentId: 1,
        unitLevel: 2,
        unitId: 201,
        childUnit: [
          {
            title: "Система управления 1.1.1",
            parentId: 201,
            unitLevel: 3,
            unitId: 251,
          },
        ],
      },
    ],
  },
  {
    title: "Система управления 2",
    unitLevel: 1,
    unitId: 2,
    childUnit: [
      {
        title: "Орган управления 2.1",
        parentId: 2,
        unitLevel: 2,
        unitId: 3,
        childUnit: [
          {
            title: "Пункт управления 2.1.1",
            parentId: 3,
            unitLevel: 3,
            unitId: 4,
            childUnit: [
              {
                title: "Подразделение 2.1.1.1",
                parentId: 4,
                unitLevel: 4,
                unitId: 5,
              },
              {
                title: "Подразделение 2.1.1.2",
                parentId: 4,
                unitLevel: 4,
                unitId: 51,
              },
            ],
          },
        ],
      },
      {
        title: "Орган управления 2.2",
        parentId: 2,
        unitLevel: 2,
        unitId: 33,
        childUnit: null,
      },
    ],
  },
];
