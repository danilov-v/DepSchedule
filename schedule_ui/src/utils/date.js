export const getNextDay = date => {
  const resultDate = new Date(
    date.getFullYear(),
    date.getMonth(),
    date.getDate()
  );
  resultDate.setDate(date.getDate() + 1);
  return resultDate;
};

export const getDates = (startDate, endDate) => {
  const results = [];
  const checkMonth = month => result => result.name === month;
  let currentDate = startDate;

  while (currentDate <= endDate) {
    const month = currentDate.toLocaleString("default", { month: "long" });
    const checkMonthFn = checkMonth(month);
    if (!results.some(checkMonthFn)) {
      results.push({ name: month, days: [] });
    }
    const currentMonth = results.find(checkMonthFn);
    currentMonth.days.push(currentDate);

    currentDate = getNextDay(currentDate);
  }

  // Improve this
  const formattedData = [...results].reverse();
  formattedData.map(month => month.days.reverse());

  return formattedData;
};

export const getAllDatesFromRange = range =>
  range.reduce((days, month) => [...days, month.days], []).flat();
