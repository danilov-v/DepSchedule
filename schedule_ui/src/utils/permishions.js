import { ROLES_PERMISSIONS } from "constants/permishions";

export const checkPermission = (role, permission) => {
  if (!role || !ROLES_PERMISSIONS[role]) {
    console.log("no such role in permissions list, chek permissions file");
    return false;
  }

  return ROLES_PERMISSIONS[role].includes(permission);
};
