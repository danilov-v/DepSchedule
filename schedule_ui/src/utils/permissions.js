import { ROLES_PERMISSIONS } from "constants/permissions";

export const checkPermission = (role, permission) => {
  if (!role || !ROLES_PERMISSIONS[role]) {
    console.log("no such role in permissions list, chek permissions file");
    return false;
  }

  return ROLES_PERMISSIONS[role].includes(permission);
};
