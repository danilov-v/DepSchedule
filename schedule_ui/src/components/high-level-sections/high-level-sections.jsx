import React, { Fragment, useEffect } from "react";
import PropTypes from "prop-types";
import { useDispatch, useSelector } from "react-redux";
import { Row } from "reactstrap";
import { cond, isEmpty, constant, stubTrue, get, isBoolean } from "lodash";
import { isWithinInterval, differenceInDays, isAfter } from "date-fns";
import { getFormattedPeriodsSelector } from "redux/selectors/periods";
import { getUserDataSelector } from "redux/selectors/ui";
import { fetchPeriods } from "redux/actions/periods";
import { HighLevelSection } from "./high-level-section";

const getDifferenceInDays = ({ day, currentSection, startDate }) =>
  differenceInDays(
    day,
    isAfter(startDate, currentSection.startDate)
      ? startDate
      : currentSection.startDate
  );

export function HighLevelSections({ range }) {
  const dispatch = useDispatch();
  const periods = useSelector(getFormattedPeriodsSelector);
  const { startDate } = useSelector(getUserDataSelector);
  const renderingSection = {};

  useEffect(() => {
    dispatch(fetchPeriods());
  }, [dispatch]);

  if (!periods.length) return null;

  return (
    <Row className="high-level-sections flex-nowrap stick-to-top" noGutters>
      {range.map(month => (
        <Fragment key={month.name}>
          {month.days.map(day => {
            const currentSection = periods.find(section =>
              isWithinInterval(day, {
                start: section.startDate,
                end: section.endDate,
              })
            );

            const renderSection = cond([
              [
                constant(isEmpty(currentSection)),
                constant(<HighLevelSection isEmpty />),
              ],
              [
                constant(
                  !isBoolean(renderingSection[get(currentSection, "name")])
                ),
                () => {
                  renderingSection[currentSection.name] = true;

                  return (
                    <HighLevelSection
                      name={currentSection.name}
                      length={
                        getDifferenceInDays({
                          day,
                          currentSection,
                          startDate,
                        }) + 1
                      }
                    />
                  );
                },
              ],
              [stubTrue, constant(null)],
            ]);

            return <Fragment key={day}>{renderSection()}</Fragment>;
          })}
        </Fragment>
      ))}
    </Row>
  );
}

HighLevelSections.propTypes = {
  range: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      days: PropTypes.arrayOf(PropTypes.object),
    })
  ),
};

HighLevelSections.defaultProps = {
  range: [],
};
