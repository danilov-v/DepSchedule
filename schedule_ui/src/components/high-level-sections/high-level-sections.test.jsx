import React from "react";
import { shallow } from "enzyme";
import { addMonths } from "date-fns";
import { getDates } from "utils/date";
import { SECTIONS } from "stub-data/sections";
import { HighLevelSections } from "./high-level-sections";

const startDate = new Date();
const endDate = addMonths(startDate, 3);

const props = {
  startDate,
  range: getDates(startDate, endDate),
  sections: SECTIONS,
};

describe("high-level-sections.jsx", () => {
  describe("<HighLevelSections />", () => {
    test("should render and match snapshot", () => {
      const component = shallow(<HighLevelSections {...props} />);

      expect(component).toMatchSnapshot();
    });
  });
});
