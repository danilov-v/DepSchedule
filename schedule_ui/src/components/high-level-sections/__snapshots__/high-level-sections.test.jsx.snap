// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`high-level-sections.jsx <HighLevelSections /> should render and match snapshot 1`] = `
<Row
  className="high-level-sections flex-nowrap"
  noGutters={true}
  tag="div"
>
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={false}
    length={5}
    name="D"
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={false}
    length={14}
    name="C"
  />
  <HighLevelSection
    isEmpty={false}
    length={10}
    name="B"
  />
  <HighLevelSection
    isEmpty={false}
    length={32}
    name="A"
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
  <HighLevelSection
    isEmpty={true}
    length={1}
    name={null}
  />
</Row>
`;
